helper/plugin/unknown.go

```diff
@@ -0,0 +1,89 @@
+package plugin
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform/configs/configschema"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// SetUnknowns takes a cty.Value, and compares it to the schema setting any null
+// leaf values which are computed as unknown.
+func SetUnknowns(val cty.Value, schema *configschema.Block) cty.Value {
+	if val.IsNull() || !val.IsKnown() {
+		return val
+	}
+
+	valMap := val.AsValueMap()
+	newVals := make(map[string]cty.Value)
+
+	for name, attr := range schema.Attributes {
+		v := valMap[name]
+
+		if attr.Computed && v.IsNull() {
+			newVals[name] = cty.UnknownVal(attr.Type)
+			continue
+		}
+
+		newVals[name] = v
+	}
+
+	for name, blockS := range schema.BlockTypes {
+		blockVal := valMap[name]
+		if blockVal.IsNull() || !blockVal.IsKnown() {
+			newVals[name] = blockVal
+			continue
+		}
+
+		blockType := blockS.Block.ImpliedType()
+
+		switch blockS.Nesting {
+		case configschema.NestingSingle:
+			newVals[name] = SetUnknowns(blockVal, &blockS.Block)
+		case configschema.NestingSet, configschema.NestingList:
+			listVals := blockVal.AsValueSlice()
+			newListVals := make([]cty.Value, 0, len(listVals))
+
+			for _, v := range listVals {
+				newListVals = append(newListVals, SetUnknowns(v, &blockS.Block))
+			}
+
+			switch blockS.Nesting {
+			case configschema.NestingSet:
+				switch len(newListVals) {
+				case 0:
+					newVals[name] = cty.SetValEmpty(blockType)
+				default:
+					newVals[name] = cty.SetVal(newListVals)
+				}
+			case configschema.NestingList:
+				switch len(newListVals) {
```

Comment:
@jbardin There's unfortunately one more wrinkle here, which you can see also in the `AllAttributesNull` function I added in an earlier commit in this PR:

If one of the attributes inside a `NestingList` block type has its type set to `cty.DynamicPseudoType` then that means each nested block object in the resulting "list" may actually have a different type, and so the result in that scenario is a `cty.TupleVal` instead of a `cty.ListVal`. A similar exception exists with `NestingMap` and `cty.ObjectVal`.

Handling that in `AllAttributesNull` was a little clunky because it is working only with the schema and not an example value, but I think handling it here is, at least, a little more straightforward: if `blockVal.Type().IsTupleType()` then we should call `cty.TupleVal` instead of going into this `switch`. The same exception applies in the `NestingMap` block if `blockVal.Type().IsObjectType()`, with the result created using `cty.ObjectVal`.

A test case for this could be to have a nested block type with an attribute of type `cty.DynamicPseudoType` and then make the input have two block objects with different types for that nested attribute. If you were to add that test case with the main code unchanged I expect you'd see a panic at the `cty.ListVal` call below, since `newListVals` would not have same-type elements.